### ISSUE TITLE
Trait produces ill-typed AIR code

### ISSUE THREAD

**Issue by parno** [verus-lang/verus#474]
The code below produces:
```
thread 'rustc' panicked at 'internal error: generated ill-typed AIR code: error 'in call to Key.lt.?, argument #2 has type "KeyInt." when it should have type "Poly"' in expression '(Key.lt.? TYPE%KeyInt. k~10$)'', rust_verify/src/verifier.rs:386:21
```
The problem vanishes if you either (a) comment out the implementation of `zero_properties` for `KeyInt`, or (b) if you remove the ensures clause from `zero_properties` in the definition of the `Key` trait.

```
#[allow(unused_imports)]
use builtin_macros::*;

#[allow(unused_imports)]
use builtin::*;

verus! {

pub trait Key : Sized {
    spec fn lt(self) -> bool;

    proof fn zero_properties()
        ensures
            forall |k:Self| k.lt();
}

struct KeyInt {
    i: usize,
}

impl Key for KeyInt {
    spec fn lt(self) -> bool { true }
    
    proof fn zero_properties() {}
}

}

fn main() { }
```

**Comment by jaybosamiya** [comment]
Just linking the _potentially_ related #370 (that one seems to require having the `recommends` clause around, while this one seems to require having the `forall` quantifier around, but otherwise basically feel like they might have the same root cause(?)

**Comment by parno** [comment]
As some additional context, here's the relevant AIR generated:
```
;; Function-Def crate::KeyInt::zero_properties<crate::Key>
(check-valid
 (axiom fuel_defaults)
 (assert
  ("postcondition not satisfied" "at the end of the function body" "failed this postcondition")
  (forall ((k~10$ KeyInt.)) (!
    (=>
     (has_type (Poly%KeyInt. k~10$) TYPE%KeyInt.)
     (%B (Key.lt.? TYPE%KeyInt. k~10$))
    )
    :pattern ((Key.lt.? TYPE%KeyInt. k~10$))
    :qid
    user_crate__KeyInt__zero_properties<crate__Key>_1
    :skolemid
    skolem_user_crate__KeyInt__zero_properties<crate__Key>_1
))))
```
It seems like either we need to be calling something like `KeyInt.lt.?`, which doesn't appear to be emitted, or that the call to `Key.lt.?` should be `(Key.lt.? TYPE%KeyInt. (Poly%KeyInt k~10$))`

**Comment by tjhance** [comment]
@Chris-Hawblitzel was this addressed by 95cecdd5542d020e36e18b22f23d241bf62ae08d?

**Comment by Chris-Hawblitzel** [comment]
> @Chris-Hawblitzel was this addressed by [95cecdd](https://github.com/verus-lang/verus/commit/95cecdd5542d020e36e18b22f23d241bf62ae08d)?

Yes.  I think this issue can be closed.
